packages/ListBox/src/store/reducer.js

```diff
@@ -0,0 +1,236 @@
+        const index = selectedOptionsArray.indexOf(activeOptionIndex);
+        selectedOptionsArray.splice(index, 1);
+      } else {
+        selectedOptionsArray.push(activeOptionIndex);
+      }
+
+      return {
+        ...state,
+        activeOption: activeOptionIndex,
+        isOpen: true,
+        onChangeFn: payload.onChangeFn,
+        selectedOptions: selectedOptionsArray,
+        shouldContentScroll: false,
+      };
+    }
+
+    case useListBox.types.hasFilter: {
+      return {
+        ...state,
+        hasFilter: payload,
+      };
+    }
+
+    case useListBox.types.applyFilter: {
+      let activeOption = null;
+      if (payload.filteredOptions.length === 1) {
+        activeOption = payload.filteredOptions[0];
+      }
+
+      if (payload.filteredOptions.length > 1) {
+        activeOption = getNextOptionActiveIndexLooping(state);
+      }
+
+      return {
+        ...state,
+        filteredOptions: payload.filteredOptions,
+        noResultsFound: payload.noResultsFound,
+        activeOption,
+      };
+    }
+
+    case useListBox.types.setTriggerWidth: {
+      return {
+        ...state,
+        triggerWidth: payload,
+      };
+    }
+
+    case useListBox.types.setHasPopupOpened: {
+      return {
+        ...state,
+        hasPopupOpened: payload,
+      };
+    }
+
+    case useListBox.types.updateOptions: {
+      const selectedOptions = Object.keys(payload)
+        .filter(key => payload[key].isSelected)
+        .map(key => Number.parseInt(key, 10));
+
+      return {
+        ...state,
+        options: payload,
+        selectedOptions,
+      };
+    }
+
+    case useListBox.types.unselectOptions: {
+      if (!Array.isArray(payload)) {
+        throw Error("unselectOptions action expect an array as a payload");
+      }
+
+      return {
+        ...state,
+        selectedOptions: state.selectedOptions.filter(index => !payload.includes(index)),
+      };
+    }
+
+    case useListBox.types.hideOptions: {
+      if (!Array.isArray(payload)) {
+        throw Error("hideOptions action expect an array as a payload");
+      }
+
+      const optionsClone = { ...state.options };
+      payload.forEach(index => {
+        optionsClone[index].isHidden = true;
+      });
+
+      return {
+        ...state,
+        options: optionsClone,
+      };
+    }
+
+    case useListBox.types.unhideOptions: {
+      if (!Array.isArray(payload)) {
+        throw Error("unhideOptions action expect an array as a payload");
+      }
+
+      const optionsClone = { ...state.options };
+      payload.forEach(index => {
+        optionsClone[index].isHidden = false;
+      });
+
+      return {
+        ...state,
+        options: optionsClone,
+      };
+    }
+
+    case useListBox.types.clear: {
+      return {
+        ...state,
+        activeOption: null,
+        isOpen: (payload && payload.isOpen) || false,
+        onChangeFn: payload && payload.onChangeFn ? payload.onChangeFn : null,
```

Comment:
Why do we need to pass around the `onChangeFn`? I'm not able to follow all the cases it's used in and it seems to be a different thing all the time.

Reply:
there are couple cases that invoke it, the onAccept clicking when has a footer, the click event and the onKeyEvent. 

At the beggining, I was manually calling the onChange in each case. But then the complexity increase and was easier to point all calls to the same function, so the user always gets the same callback answer no matter which event is occurring. 

So for the user is transparent no matter how the user interacts the consumer will always receive and nice structure answer in the onChange event.